alf/networks/encoding_networks.py

```diff
@@ -0,0 +1,304 @@
+        self._start_decoding_shape = [
+            start_decoding_channels, start_decoding_height,
+            start_decoding_width
+        ]
+        self._preprocess_fc_layers.append(
+            layers.FC(
+                input_size,
+                np.prod(self._start_decoding_shape),
+                activation=activation))
+
+        self._transconv_layer_params = transconv_layer_params
+        self._transconv_layers = nn.ModuleList()
+        in_channels = start_decoding_channels
+        for i, paras in enumerate(transconv_layer_params):
+            filters, kernel_size, strides = paras[:3]
+            padding = paras[3] if len(paras) > 3 else 0
+            act = activation
+            if i == len(transconv_layer_params) - 1:
+                act = output_activation
+            self._transconv_layers.append(
+                layers.ConvTranspose2D(
+                    in_channels,
+                    filters,
+                    kernel_size,
+                    activation=act,
+                    strides=strides,
+                    padding=padding))
+            in_channels = filters
+
+    def output_shape(self):
+        """Return the output image shape given the start_decoding_shape.
+
+        How to calculate the output size:
+        https://pytorch.org/docs/stable/nn.html#torch.nn.ConvTranspose2d
+
+            H = (H1-1) * strides + HF - 2P
+
+        where H = output size, H1 = input size, HF = size of kernel, P = padding
+
+        Returns:
+            a tuple representing the output shape (C,H,W)
+        """
+        _, height, width = self._start_decoding_shape
+        for paras in self._transconv_layer_params:
+            filters, kernel_size, strides = paras[:3]
+            padding = paras[3] if len(paras) > 3 else 0
+            height = (height - 1) * strides + kernel_size - 2 * padding
+            width = (width - 1) * strides + kernel_size - 2 * padding
+        return (filters, height, width)
+
+    def forward(self, inputs):
+        """Returns an image of shape (B,C,H,W)."""
+        assert len(inputs.size()) == 2, \
+            "The input dims {} are incorrect! Should be (B,N)".format(
+                inputs.size())
+        z = inputs
+        for fc_l in self._preprocess_fc_layers:
+            z = fc_l(z)
+        z = z.view(-1, *self._start_decoding_shape)
+        for deconv_l in self._transconv_layers:
+            z = deconv_l(z)
+        return z
+
+
+class EncodingNetwork(nn.Module):
+    """Feed Forward network with CNN and FC layers."""
+
+    def __init__(self,
+                 input_tensor_spec,
+                 conv_layer_params=None,
+                 fc_layer_params=None,
+                 activation=layers.identity,
+                 last_layer_size=None,
+                 last_activation=None):
+        """Create an EncodingNetwork
+
+        This EncodingNetwork allows the last layer to have different settings
+        from the other layers.
+
+        Args:
```

Comment:
TODO: Handle nested input_tensor_spec

Reply:
I don't think it's necessary to complicate the APIs in general. The users can write subclasses for specific problems. For now, unnested tensor specs is the common case.